File: alert-platform/services/delivery_trueconf/templates.py
"""Шаблоны уведомлений — раздел 9.3: конверт, а не письмо. Тело — дословный
оригинал источника (И2), не проходит через шаблонизацию. Шапка и подвал —
шаблон.

Сознательно урезано относительно раздела 9.3 для M4: строка действий
("Ответьте: взял · не мой · следствие") не включена — полная семантика
раздела 9.4 не реализована (см. packages/common/ack.py — только бинарный
факт реакции, минимальный срез). Ссылка в личный кабинет и на текущие
алерты (Этап 4) — реализованы, включаются в render_new ниже.
"""
from __future__ import annotations

from datetime import datetime

PRIORITY_EMOJI = {"P0": "🔴", "P1": "🟠", "P2": "🟡", "P3": "⚪"}


def display_id(problem_id: int, incident_id: int | None) -> str:
    return f"INC-{incident_id:04d}" if incident_id else f"PRB-{problem_id:04d}"


def render_new(*, problem_id: int, incident_id: int | None, priority: str | None,
                object_name: str, site_name: str, service_name: str | None,
                source_system: str, original_body: str, symptom_class: str,
                ai_root_cause_hypothesis: str | None = None, alerts_link: str | None = None) -> str:
    emoji = PRIORITY_EMOJI.get(priority or "P3", "⚪")
    did = display_id(problem_id, incident_id)
    service_line = f"Сервис: {service_name}" if service_name else "Сервис: не определён"
    text = (
        f"{emoji} <b>{priority or '?'} · {did}</b> · {object_name} · {site_name}\n"
        f"─────── оригинал {source_system} ───────\n"
        f"{original_body}\n"
        f"───────────────────────────────\n"
        f"{service_line} · Симптом: {symptom_class}"
    )
    if ai_root_cause_hypothesis:
        # раздел 6.1/13 — коррелятор не связал событие правилом (например,
        # раздел 18.4 site_outage_ambiguous), ИИ-гипотеза дополняет, но не
        # заменяет факты выше; явно маркирована как предположение.
        text += (f"\n\n<i>Вероятная первопричина (гипотеза, требует проверки, "
                 f"сформирована ИИ):</i>\n{ai_root_cause_hypothesis}")
    if alerts_link:
        # Этап 4 — сотрудник должен суметь одним тапом из самого алерта
        # увидеть полную текущую картину и последовательность (не только
        # этот один алерт), не набирая команду /алерты вручную.
        text += f"\n\nВаши текущие алерты и последовательность: {alerts_link}"
    return text


def render_closure(*, problem_id: int, incident_id: int | None,
                    resolved_at: datetime, duration_text: str,
                    closed_by_reconciliation: bool) -> str:
    did = display_id(problem_id, incident_id)
    note = " (закрыто автоматически по таймауту, без сообщения о восстановлении)" \
        if closed_by_reconciliation else ""
    return (
        f"🟢 <b>ЗАКРЫТО · {did}</b>\n"
        f"Восстановлено {resolved_at:%Y-%m-%d %H:%M:%S} · Длительность {duration_text}{note}"
    )


def render_supplement(*, problem_id: int, incident_id: int, root_object: str,
                       root_symptom_class: str, opened_at: datetime, n_symptoms: int,
                       n_services: int, rule_names: list[str], ai_summary: str | None,
                       ai_recommendation: str | None = None,
                       checklist_fallback: list[str] | None = None) -> str:
    did = display_id(problem_id, incident_id)
    rules = ", ".join(rule_names) or "не определено"
    lines = [
        f"🔵 <b>ДОПОЛНЕНИЕ к {did}</b>",
        f"Первопричина: {root_object} ({root_symptom_class}), с {opened_at:%Y-%m-%d %H:%M:%S}",
        f"Связано алертов: {n_symptoms} · Затронуто сервисов: {n_services}",
        f"Основание: правило {rules}",
    ]
    if ai_summary:
        # раздел 13: вывод модели помечен как гипотеза, ниже фактов правила.
        lines.append("")
        lines.append("<i>Сводка (гипотеза, сформирована ИИ):</i>")
        lines.append(ai_summary)
    if ai_recommendation:
        # ИИ выбрала и сформулировала пункты из чек-листа (packages/ai/recommend.py)
        # под конкретику инцидента — сами пункты не выдуманы, только фразировка/отбор.
        lines.append("")
        lines.append("<i>Рекомендация (на основе базы знаний, сформулирована ИИ):</i>")
        lines.append(ai_recommendation)
    elif checklist_fallback:
        # раздел И5: ИИ недоступна — показываем чек-лист как есть, без фразировки.
        lines.append("")
        lines.append("<i>Рекомендация (чек-лист из базы знаний):</i>")
        lines.extend(f"• {step}" for step in checklist_fallback)
    return "\n".join(lines)


def render_duplicate_note(*, duplicate_problem_id: int, original_problem_id: int,
                           original_incident_id: int | None, source_system: str) -> str:
    """Раздел 4.1 — ИИ определила, что это тот же реальный объект,
    описанный другим источником другими словами (packages/ai/dedup.py).
    Короткая пометка вместо ещё одного полноценного NEW — раздел И2/И3
    не нарушены: оригинал всё ещё доступен по PRB-id для трассировки."""
    did = display_id(original_problem_id, original_incident_id)
    return (
        f"🔗 <b>ДУБЛЬ</b> · PRB-{duplicate_problem_id:04d}\n"
        f"Похоже на то же событие, что и {did} — подтверждение от {source_system} "
        f"(определено ИИ, раздел 4.1). Отдельное уведомление не отправлено."
    )


def render_scenario_notify(*, problem_id: int, incident_id: int | None, scenario_name: str,
                            object_name: str, is_escalation: bool) -> str:
    """Раздел «Сценарии», Этап 2 — уведомление по шагу линейной цепочки
    (packages/scenarios/engine.py). is_escalation=True — это шаг ПОСЛЕ
    «Подождать» (дедлайн истёк раньше, чем проблема решилась), не первое
    уведомление сценария."""
    did = display_id(problem_id, incident_id)
    kind = "ЭСКАЛАЦИЯ ПО СЦЕНАРИЮ" if is_escalation else "СЦЕНАРИЙ"
    return (
        f"🟣 <b>{kind}: {scenario_name}</b> · {did}\n"
        f"Объект: {object_name}"
    )


def render_sla_breach(*, problem_id: int, incident_id: int | None, object_name: str,
                       priority: str | None, age_minutes: int, threshold_minutes: int,
                       rule_name: str) -> str:
    """Раздел «SLA», Этап 2 — напоминание, что проблема в работе дольше
    норматива реагирования правила. Факты (сколько прошло, какой порог),
    не домысел — тот же принцип, что и в остальных шаблонах файла."""
    did = display_id(problem_id, incident_id)
    return (
        f"⏰ <b>НАРУШЕН SLA · {did}</b>\n"
        f"{object_name} · приоритет {priority or '?'}\n"
        f"В работе {age_minutes} мин · порог реагирования по правилу «{rule_name}» — {threshold_minutes} мин"
    )


def format_duration(seconds: float) -> str:
    seconds = int(seconds)
    h, rem = divmod(seconds, 3600)
    m, s = divmod(rem, 60)
    if h:
        return f"{h} ч {m} мин"
    if m:
        return f"{m} мин {s} с"
    return f"{s} с"

File: alert-platform/go-platform/internal/planner/templates.go
package planner

import (
	"fmt"
	"html"
	"strings"
	"time"
)

// esc экранирует динамический текст (сырое тело алерта от источника
// мониторинга, вывод ИИ, имена объектов/сценариев/правил) перед
// вставкой в HTML-размеченное сообщение TrueConf (parse_mode: HTML).
// Не применяется к литеральной разметке (<b>/<i>), которую пишет сам
// шаблон — только к значениям из данных. Без этого скомпрометированный
// или ошибочно настроенный источник мониторинга мог бы протащить
// произвольную HTML-разметку в исходящее сообщение через текст алерта.
func esc(value string) string {
	return html.EscapeString(value)
}

var priorityEmoji = map[string]string{
	"P0": "🔴",
	"P1": "🟠",
	"P2": "🟡",
	"P3": "⚪",
}

func DisplayID(problemID int64, incidentID *int64) string {
	if incidentID != nil {
		return fmt.Sprintf("INC-%04d", *incidentID)
	}
	return fmt.Sprintf("PRB-%04d", problemID)
}

func RenderNew(data ProblemData) string {
	priority := valueOr(data.Priority, "?")
	emoji, ok := priorityEmoji[priority]
	if !ok {
		emoji = "⚪"
	}
	serviceLine := "Сервис: не определён"
	if data.ServiceName != nil {
		serviceLine = "Сервис: " + esc(*data.ServiceName)
	}
	text := fmt.Sprintf(
		"%s <b>%s · %s</b> · %s · %s\n─────── оригинал %s ───────\n%s\n───────────────────────────────\n%s · Симптом: %s",
		emoji, priority, DisplayID(data.ID, data.IncidentID), esc(data.ObjectName),
		esc(valueOr(data.Site, "?")), esc(data.SourceSystem), esc(data.OriginalBody),
		serviceLine, esc(data.SymptomClass),
	)
	if data.AIRootCauseHypothesis != nil && *data.AIRootCauseHypothesis != "" {
		text += "\n\n<i>Вероятная первопричина (гипотеза, требует проверки, сформирована ИИ):</i>\n" +
			esc(*data.AIRootCauseHypothesis)
	}
	return text
}

func RenderClosure(data ProblemData) string {
	note := ""
	if data.ClosedByReconciliation {
		note = " (закрыто автоматически по таймауту, без сообщения о восстановлении)"
	}
	return fmt.Sprintf(
		"🟢 <b>ЗАКРЫТО · %s</b>\nВосстановлено %s · Длительность %s%s",
		DisplayID(data.ID, data.IncidentID),
		data.ResolvedAt.Format("2006-01-02 15:04:05"),
		FormatDuration(data.ResolvedAt.Sub(data.OpenedAt)), note,
	)
}

func RenderDuplicate(duplicateProblemID, originalProblemID int64, incidentID *int64, source string) string {
	return fmt.Sprintf(
		"🔗 <b>ДУБЛЬ</b> · PRB-%04d\nПохоже на то же событие, что и %s — подтверждение от %s (определено ИИ, раздел 4.1). Отдельное уведомление не отправлено.",
		duplicateProblemID, DisplayID(originalProblemID, incidentID), esc(source),
	)
}

func RenderScenario(problem ProblemData, scenarioName string, escalation bool) string {
	action := "Уведомление"
	if escalation {
		action = "Эскалация"
	}
	return fmt.Sprintf("🧩 <b>%s · %s</b>\nСценарий: %s\nОбъект: %s · Приоритет: %s", action, DisplayID(problem.ID, problem.IncidentID), esc(scenarioName), esc(problem.ObjectName), valueOr(problem.Priority, "?"))
}

func RenderSLABreach(problem ProblemData, ruleName string, ageMinutes, thresholdMinutes int) string {
	return fmt.Sprintf("⏱ <b>НАРУШЕНИЕ SLA · %s</b>\nОбъект: %s · Приоритет: %s\nНет реакции %d мин (порог %d мин) · правило %s", DisplayID(problem.ID, problem.IncidentID), esc(problem.ObjectName), valueOr(problem.Priority, "?"), ageMinutes, thresholdMinutes, esc(ruleName))
}

type SupplementData struct {
	ProblemID        int64
	IncidentID       int64
	RootObject       string
	RootSymptom      string
	OpenedAt         time.Time
	SymptomsCount    int
	ServicesCount    int
	RuleNames        []string
	AISummary        *string
	AIRecommendation *string
	Checklist        []string
}

func RenderSupplement(data SupplementData) string {
	rules := "не определено"
	if len(data.RuleNames) > 0 {
		rules = esc(strings.Join(data.RuleNames, ", "))
	}
	lines := []string{
		fmt.Sprintf("🔵 <b>ДОПОЛНЕНИЕ к %s</b>", DisplayID(data.ProblemID, &data.IncidentID)),
		fmt.Sprintf("Первопричина: %s (%s), с %s", esc(data.RootObject), esc(data.RootSymptom), data.OpenedAt.Format("2006-01-02 15:04:05")),
		fmt.Sprintf("Связано алертов: %d · Затронуто сервисов: %d", data.SymptomsCount, data.ServicesCount),
		fmt.Sprintf("Основание: правило %s", rules),
	}
	if data.AISummary != nil && *data.AISummary != "" {
		lines = append(lines, "", "<i>Сводка (гипотеза, сформирована ИИ):</i>", esc(*data.AISummary))
	}
	if data.AIRecommendation != nil && *data.AIRecommendation != "" {
		lines = append(lines, "", "<i>Рекомендация (на основе базы знаний, сформулирована ИИ):</i>", esc(*data.AIRecommendation))
	} else if len(data.Checklist) > 0 {
		lines = append(lines, "", "<i>Рекомендация (чек-лист из базы знаний):</i>")
		for _, step := range data.Checklist {
			lines = append(lines, "• "+esc(step))
		}
	}
	return strings.Join(lines, "\n")
}

type AIAnalysisData struct {
	ProblemID    int64
	IncidentID   *int64
	ObjectName   string
	SymptomClass string
	Site         *string
	Priority     *string
	RelatedCount int
	AIText       *string
}

func RenderAIAnalysis(data AIAnalysisData) string {
	lines := []string{
		fmt.Sprintf("🧠 <b>РАЗБОР ПО ЗАПРОСУ · %s</b>", DisplayID(data.ProblemID, data.IncidentID)),
		fmt.Sprintf("Объект: %s · Симптом: %s · Площадка: %s · Приоритет: %s",
			esc(data.ObjectName), esc(data.SymptomClass), esc(valueOr(data.Site, "?")), valueOr(data.Priority, "?")),
	}
	if data.IncidentID != nil {
		lines = append(lines, fmt.Sprintf("Связанных алертов в инциденте: %d", data.RelatedCount))
	} else {
		lines = append(lines, "Инцидент не сформирован — разбор по одиночному алерту")
	}
	if data.AIText != nil && *data.AIText != "" {
		lines = append(lines, "", "<i>Разбор (гипотеза, требует проверки, сформирована ИИ):</i>", esc(*data.AIText))
	} else {
		lines = append(lines, "", "ИИ временно недоступна — попробуйте попросить разбор ещё раз чуть позже.")
	}
	return strings.Join(lines, "\n")
}

func FormatDuration(duration time.Duration) string {
	seconds := int64(duration.Seconds())
	hours := seconds / 3600
	minutes := (seconds % 3600) / 60
	remainder := seconds % 60
	if hours > 0 {
		return fmt.Sprintf("%d ч %d мин", hours, minutes)
	}
	if minutes > 0 {
		return fmt.Sprintf("%d мин %d с", minutes, remainder)
	}
	return fmt.Sprintf("%d с", remainder)
}

func valueOr(value *string, fallback string) string {
	if value == nil || *value == "" {
		return fallback
	}
	return *value
}
